Fix library agnosticism in random locations

tylerriccio33 opened the pull request:
This is meant to fix #675 in GT and https://github.com/posit-dev/pointblank/issues/132 in pointblank.

I'm brute forcing it since I see no other way. I'm replacing the single dispatch paradigm (in effected places) with an attempt to import the library in question.

FYI I'm 99% sure this would be solved by Narwhals if you wanted to go that route down the line.

tylerriccio33 commented:
Actually, can I get another set of eyes on this issue? The scope all of the sudden seems massive. I thought it was contained to a few accidental instance checks on single dispatch methods, but it appears to be instance checking a TON of functions.

If I'm scoping this correctly, I think there are 3 options:

1. Doing a naive check like 'pandas' in str(type(df)): ...
2. Doing a try-except import check like try: __import__('pandas') ...
3. Adopting narwhals

I'm currently unaware of a way to do an instance check on a class you do not have available in your process. Happy to help out w/any of these, I would be aided by this being solved.

I'd be even happier if you were to tell me I'm wrong and the problem is far more contained :)

machow commented:
Hey, thanks for taking a look at this, and raising the original issue. If you have the chance, do you mind taking a look at this explanation, and letting me know if it does what you need?

The tl;dr is that you should be able to pass a polars series to the val functions, but we should better document this 😅 

https://github.com/posit-dev/great-tables/issues/675#issuecomment-2855706009

jrycw commented:
@tylerriccio33, You might find our [Bring Your Own DataFrame](https://posit-dev.github.io/great-tables/blog/bring-your-own-df/#how-we-made-pandas-optional) blog post helpful—feel free to check it out.

tylerriccio33 commented:
I'm not the expert in how single dispatch decorator works in python, but it appears to be running an is instance check when to_frame is called, which I don't think is intended. Even when i add code to convert the the list to a polars series, then pass it to to_frame, it does an instance check and fails since pandas isn't available. 

I feel I may be not understanding a key concept here, and there can't fix this. I'd really like to fix this bug, I'm running into some compatibility issues at work; it's sometimes difficult to load prod containers w/pandas due to other compatibility concerns (arrow/numpy). Let me know if i'm understanding this correctly.

Pushed a new pr to show the test I'm using, which may also be the problem?

jrycw commented:
I came across a [PyCon 2025 talk](https://www.youtube.com/watch?v=hidy15rK2a4) on YouTube that briefly yet clearly outlines four different ways to choose actions based on instance types.

tylerriccio33 commented:
The error is happening before the function code is called, it's happening during the evaluation process in singledispatch. In `AbstractBackend` `__subclasshook__` method, I pass a series but the only cls._backend is a Pandas series, so it cannot import the module. This feels very specific to the data backend implementation so I'm having a hard time debugging where the backend class attribute is set...

tylerriccio33 commented:
This pr is imperfect but working. I found it wasn't working because Polar's `Series` object isn't defined or available at the top level of the module. Instead, it's down a few levels. To get around this I added the full path (series.series.Series) to the backend.

`class PlSeries(AbstractBackend):
        _backends: list[tuple[str, str]] = [("polars", "series.series.Series")]`
        
I then added additional logic to allow for a class name not defined at the top level:

`class AbstractBackend(metaclass=_AbstractBackendMeta):
    @classmethod
    def __init_subclass__(cls):
        if not hasattr(cls, "_backends"):
            cls._backends = []

    @classmethod
    def __subclasshook__(cls, subclass):
        for mod_name, cls_name in cls._backends:
            try:
                parent_candidate = _load_class(mod_name, cls_name)
                if issubclass(subclass, parent_candidate):
                    return True
            except (ModuleNotFoundError, ImportError):
                continue

        return NotImplemented`
        
Again, not the most elegant but it appears to be effective. I've struggled to wrap my head around some aspects of how the backend works, so it's possible I'm missing finer points.

Also, strictly for development purposes, I redefined some dev deps in the pyproject.toml file for uv to find, and added uv.lock to the gitignore. Feel free to reject that, just helped me code faster!
